### PR TITLE
[kotlin-racingcar-precourse] Submit ahmed's mission

### PR DESCRIPTION
**Pull request by ahmedelshaikh20**
- I implemented a simple car racing game using Kotlin as specified in the pre-course mission. I structured the project using proper packages (model, core, parser, utils) to separate concerns clearly.
 
- I focused on writing clean, testable code that follows Kotlin conventions. I used the Car class to model each car's state, a Game class to manage the race, and utility classes for input parsing, random logic, and printing results.
 
- I added thorough input validation to handle invalid car names and round counts. I also ensured that output formatting matched the mission's expectations exactly.

- Additionally, I wrote unit tests to verify the behavior of the core classes (Car, Game, and Parser). All tests pass, and the game behaves correctly even in edge cases like zero rounds or tied winners.
 
- Throughout the project, I committed in small, meaningful steps using Angular-style commit messages. I completed the README checklist and reflected on the challenges I faced.
 
- Although I could have **overengineered** the project by introducing more abstraction (like using strategy patterns or dependency injection), I chose to keep it simple and readable to match the mission's scope. The current structure is easy to understand, extend, and maintain. Overall, I believe the mission was easy to follow and done.
 

**Thanks in advance for reviewing my submission. I'm excited to keep learning and improving and truly looking forward to the opportunity of joining the bootcamp.**

### CODE REVIEW

**Comment by jinnyjiinlee** [comment]
Hello 👋 I'm Jiin Lee from South Korea :)
I really enjoyed checking out your code — I'd love to do a code review together if you're interested too!
Feel free to stop by my PR and leave a comment. I'd really appreciate it!

[My PR Address]
https://github.com/herotechcourse/kotlin-racingcar-precourse/pull/70

File: src/main/kotlin/racingcar/Application.kt [hunk 1]
```diff
@@ -1,5 +1,22 @@
...
 fun main() {
-    // TODO: Implement the program
+
+
+  val carsString = InputHandler.readCarNames()
+  val rounds = InputHandler.readRoundCount()
+  val parsedCars = parseCars(carsString)
+  val game = Game(parsedCars, rounds)
+
+  println("Race Results:")
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/Application.kt:16]: 'HeroTechCourse' mentioned to keep the string constant, so I think is would be better to make this a constant as well.

File: src/main/kotlin/racingcar/utils/InputHandler.kt [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package racingcar.utils
+
+import camp.nextstep.edu.missionutils.Console
+
+object InputHandler {
+
+  fun readCarNames(): String {
+    println("Enter car names (comma-separated): ")
+    return Console.readLine()
+  }
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/utils/InputHandler.kt:10]: Is there a reason why the constants weren't separated in this part of the code as well?

File: src/test/kotlin/racingcar/ParserTest.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+
+class ParserTest {
+
+  @Test
+  fun `parses valid car names into list of Car`() {
+    val input = "audi,ford,bmw"
+    val cars = Parser.parseCars(input)
+
+    assertThat(cars).hasSize(3)
```
Comment (jinnyjiinlee) [inline src/test/kotlin/racingcar/ParserTest.kt:18]: The requirements mentioned using AssertJ, but I didn't in my RacingCar Mission. 
You did a great job using it! I also think using hasSize to check the length of cars was a really nice touch!

File: src/main/kotlin/racingcar/Application.kt [hunk 4]
```diff
@@ -1,5 +1,22 @@
...
 
+import racingcar.core.Game
+import racingcar.parser.Parser.parseCars
+import racingcar.utils.InputHandler
+import racingcar.utils.Printer.printWinners
+
 fun main() {
-    // TODO: Implement the program
+
+
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/Application.kt:10]: I was curious if there was a particlular reason for adding two line breaks here?

File: src/main/kotlin/racingcar/Application.kt [hunk 5]
```diff
@@ -1,5 +1,22 @@
...
+  val carsString = InputHandler.readCarNames()
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/Application.kt:11]: Since HeroTechCourse suggests avoiding types in variable names, I think it might be a good idea to rename this.

File: src/main/kotlin/racingcar/utils/InputHandler.kt [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package racingcar.utils
+
+import camp.nextstep.edu.missionutils.Console
+
+object InputHandler {
+
+  fun readCarNames(): String {
+    println("Enter car names (comma-separated): ")
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/utils/InputHandler.kt:8]: What do you think about extracting this as a constant? (based on the HoroTechCourse requirements)

File: src/main/kotlin/racingcar/parser/Parser.kt [hunk 7]
```diff
@@ -0,0 +1,25 @@
+package racingcar.parser
+
+import racingcar.utils.ErrorMessages
+import racingcar.model.Car
+
+object Parser {
+  fun parseCars(input: String): MutableList<Car> {
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/parser/Parser.kt:7]: The object is already called 'Parser', so I was thinking that nameing the function simplay 'car' might be enough to get the idea acroos.

File: src/main/kotlin/racingcar/parser/Parser.kt [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+    val names = input
+      .split(",")
+      .map { it.trim() }
+
+    require(names.size == names.distinct().size) {
+      ErrorMessages.ERROR_DUPLICATE_CAR_NAME
+    }
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/parser/Parser.kt:15]: This is a'require' method I learned about recently - I think using 'require' for exception handling really improves readability!

File: src/main/kotlin/racingcar/parser/Parser.kt [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+    return names
+      .onEach {
+        require(it.isNotEmpty()) { ErrorMessages.ERROR_EMPTY_CAR_NAME }
+        require(it.length <= 5) { ErrorMessages.ERROR_LONG_CAR_NAME }
+      }
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/parser/Parser.kt:21]: You used the `onEach` method - it's great for validating each element!
Thanks! I learned something new from you - I didn't know about this method before.

File: src/main/kotlin/racingcar/core/Game.kt [hunk 10]
```diff
@@ -0,0 +1,25 @@
+package racingcar.core
+
+import racingcar.utils.Printer.printRound
+import racingcar.model.Car
+
+class Game(private val cars: List<Car>, private val rounds: Int) {
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/core/Game.kt:6]: These functions in the `Game` class are mostily short and focused on a single responsibility, which makes the code very readable :) 👍

File: src/main/kotlin/racingcar/utils/Utils.kt [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package racingcar.utils
+
+import camp.nextstep.edu.missionutils.Randoms
+
+object Utils {
+  fun getRandomNumberFromZeroToNine(): Int {
+    return Randoms.pickNumberInRange(0, 9)
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/utils/Utils.kt:7]: Since 0 and 9 represent meaningful bounds here, mayble consider defining them as constatns instead of using magin number.

File: src/main/kotlin/racingcar/model/Car.kt [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+  fun getProgress(): String {
+    return "-".repeat(currLevel)
+  }
+
+  fun tryMove() {
+    if (getRandomNumberFromZeroToNine() >= MOVE_THRESHOLD) move()
+  }
+
+}
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/model/Car.kt:24]: You managed the `Car`logic really well by making it into an object.
I think it is great!

File: src/main/kotlin/racingcar/Application.kt [hunk 13]
```diff
@@ -1,5 +1,22 @@
...
+  val rounds = InputHandler.readRoundCount()
+  val parsedCars = parseCars(carsString)
+  val game = Game(parsedCars, rounds)
+
+  println("Race Results:")
+  game.startRace()
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/Application.kt:17]: I was struggling to come up with a good constant name for '---', but I think the one you chose - `startRace()` - is a great fit!

File: src/main/kotlin/racingcar/model/Car.kt [hunk 14]
```diff
@@ -0,0 +1,24 @@
+package racingcar.model
+
+import racingcar.utils.Utils.getRandomNumberFromZeroToNine
+
+
+data class Car(val name: String, var currLevel: Int = 0) {
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/model/Car.kt:6]: This is a minor point, but I found `currLevel` a bit unclear at first glance.
Maybe using a more intuitive name like `position`could make it easier to understand - just a personal thought!

File: src/main/kotlin/racingcar/utils/ErrorMessages.kt [hunk 15]
```diff
@@ -0,0 +1,9 @@
+package racingcar.utils
+
+object ErrorMessages {
+  const val ERROR_EMPTY_CAR_NAME = "Car names must not be empty."
+  const val ERROR_LONG_CAR_NAME = "Car names must be 5 characters or less."
+  const val ERROR_INVALID_ROUNDS = "Rounds must be a valid positive integer."
+  const val ERROR_DUPLICATE_CAR_NAME = "Car names must be unique."
+
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/utils/ErrorMessages.kt:8]: Looks like this line break was added unintentioanlly~

File: src/main/kotlin/racingcar/utils/ErrorMessages.kt [hunk 16]
```diff
@@ -0,0 +1,9 @@
+package racingcar.utils
+
+object ErrorMessages {
+  const val ERROR_EMPTY_CAR_NAME = "Car names must not be empty."
+  const val ERROR_LONG_CAR_NAME = "Car names must be 5 characters or less."
+  const val ERROR_INVALID_ROUNDS = "Rounds must be a valid positive integer."
+  const val ERROR_DUPLICATE_CAR_NAME = "Car names must be unique."
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/utils/ErrorMessages.kt:7]: Nice work on the thorough validation checks!

File: src/main/kotlin/racingcar/utils/Printer.kt [hunk 17]
```diff
@@ -0,0 +1,17 @@
+package racingcar.utils
+
+import racingcar.model.Car
+
+object Printer {
+
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/utils/Printer.kt:6]: Was this line break added on purpose?

File: src/test/kotlin/racingcar/ParserTest.kt [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  @Test
+  fun `throws exception if any car name exceeds 5 characters`() {
+    val input = "audi,chevrolet,bmw"
+
+    assertThatThrownBy { Parser.parseCars(input) }
+      .isInstanceOf(IllegalArgumentException::class.java)
+      .hasMessage(ErrorMessages.ERROR_LONG_CAR_NAME)
+  }
```
Comment (jinnyjiinlee) [inline src/test/kotlin/racingcar/ParserTest.kt:46]: I think this test could be a good place to use a @ParameterizedTest, especially if you plan to check multiple invalid inputs, It might make the test more scalable and concise!

File: src/main/kotlin/racingcar/utils/InputHandler.kt [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+    return Console.readLine()
+  }
+
+  fun readRoundCount(): Int {
+    println("How many rounds will be played?")
```
Comment (jinnyjiinlee) [inline src/main/kotlin/racingcar/utils/InputHandler.kt:13]: What do you think about extracting this as a contant?

**Review by jinnyjiinlee — COMMENTED** [review]
Hey Ahmed~  :) 
I really enjoyed reading through your code.
I could clearly see the thought and effort you put into every part of it.

From what I've seen, HeroTechCourse seems to value readability and simplicity a lot, so I tried to focus my feedback around that.

I hope my comments are helpful to you!
If anything I mentioned is unclear, feel free to replay ^^!

Great job over the past two weeks, and well done on completing the mission 💪✨
